File: python/longest_substring.py
from curses import longname


def lengthOfLongestSubstring(s: str) -> int:
        if s == '': return 0
        c_map = {} # Keep the last index of characters
        p = -1 # pointer to index we start to count from
        max_length = 0
        for i in range(len(s)):
            c = s[i]
        # if c is in the current substring
            if c in c_map and p < c_map[c]:
        # update the pointer
                p = c_map[c]
            else:
        # else, update the max length
                max_length = i - p if max_length < i-p else max_length
        # update the last index of c
            c_map[c] = i
            
        return max_length

print(lengthOfLongestSubstring('pwwkew'))

File: javascript/longestSubString.js
// leetcode.com/problems/longest-substring-without-repeating-characters/
https: var lengthOfLongestSubstring = function (s) {
	if (s === '') return 0;
	cMap = {};
	p = -1;
	max = 0;
	for (let i = 0; i < s.length; i++) {
		c = s[i];
		if (cMap[c] !== undefined && p < cMap[c]) {
			// if c in the current substring, update the pointer
			p = cMap[c];
		} else {
			// else, update max
			max = max < i - p ? i - p : max;
		}
		cMap[c] = i;
	}

	return max;
};

console.log(lengthOfLongestSubstring(''));
